cscs-checks/apps/greasy/greasy_check.py

```diff
@@ -0,0 +1,218 @@
+            self.variables['CRAY_CUDA_MPS'] = '1'
+            self.variables['CUDA_VISIBLE_DEVICES'] = '0'
+            self.variables['GPU_DEVICE_ORDINAL'] = '0'
+            self.extra_resources = {
+                'gres': {
+                    'gres': 'gpu:0,craynetwork:4'
+                }
+            }
+        elif self.current_partition.fullname in ['daint:mc', 'dom:mc']:
+            self.extra_resources = {
+                'gres': {
+                    'gres': 'craynetwork:72'
+                }
+            }
+
+    @rfm.run_before('run')
+    def change_executable_name(self):
+        # After compiling the code we can change the executable to be
+        # the greasy one
+        self.executable = 'greasy'
+
+    @rfm.run_before('run')
+    def set_launcher(self):
+        # The job launcher has to be changed to local since greasy
+        # make calls to srun
+        self.job.launcher = getlauncher('local')()
+
+    @sn.sanity_function
+    def eval_sanity(self):
+        output_files = []
+        output_files = [file for file in os.listdir(self.stagedir)
+                                      if file.startswith('output-')]
+        num_greasy_tasks = len(output_files)
+        failure_msg = (f'Requested {self.num_greasy_tasks} task(s), but '
+                       f'executed only {num_greasy_tasks} tasks(s)')
+        sn.evaluate(sn.assert_eq(num_greasy_tasks, self.num_greasy_tasks,
+                                 msg=failure_msg))
+        num_tasks = sn.getattr(self, 'nranks_per_worker')
+        num_cpus_per_task = sn.getattr(self, 'num_cpus_per_task')
+
+        def tid(match):
+            return int(match.group(1))
+
+        def num_threads(match):
+            return int(match.group(2))
+
+        def rank(match):
+            return int(match.group(3))
+
+        def num_ranks(match):
+            return int(match.group(4))
+
+        for output_file in output_files:
+            result = sn.findall(r'Hello, World from thread \s*(\d+) out '
+                                r'of \s*(\d+) from process \s*(\d+) out of '
+                                r'\s*(\d+)', output_file)
+
+            failure_msg = (f'Found {sn.count(result)} Hello, World... '
+                           f'pattern(s) but expected '
+                           f'{num_tasks * num_cpus_per_task} pattern(s) inside '
+                           f'the output file {output_file}')
+            sn.evaluate(sn.assert_eq(sn.count(result),
+                                     num_tasks * num_cpus_per_task,
+                                     msg=failure_msg))
+
+            sn.evaluate(sn.all(
+                sn.chain(
+                    sn.map(lambda x: sn.assert_lt(tid(x), num_threads(x),
+                                                  msg=f'Found {tid(x)} threads '
+                                                  'rather than '
+                                                  f'{num_threads(x)}'), result),
```

Comment:
Better write all those as follows, so that PEP8 is happy:

```suggestion
                    sn.map(lambda x: sn.assert_lt(
                        tid(x), num_threads(x),
                        msg=(f'Found {tid(x)} threads rather than '
                             f'{num_threads(x)}'), result
                    ),
```